### DISCUSSION TITLE
"CPU not defined" error

### DISCUSSION THREAD

**Discussion by Lichriszz** [CliMA/Oceananigans.jl#1362] [Computational science!]
Hi,

When I run the shallow water model, I got an error message saying:

"LoadError: UndefVarError: CPU not defined
in expression starting at /Users/.../Desktop/Research/shallow water test.jl:8
(::Core.var"#Type##kw")(::NamedTuple{(:grid, :gravitational_acceleration),Tuple{RegularCartesianGrid{Float64,Periodic,Periodic,Bounded,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},Nothing}}, ::Type{ShallowWaterModel}) at shallow_water_model.jl:75
top-level scope at shallow water test.jl:8
include_string(::Function, ::Module, ::String, ::String) at loading.jl:1088"

Here are my codes:

using Pkg
pkg"add Oceananigans, JLD2, Plots"
using Oceananigans, Oceananigans.Advection, Oceananigans.Architectures

grid = RegularCartesianGrid(size=(128, 128, 1), extent=(2π, 2π, 2π))

model = ShallowWaterModel(grid = grid,
                          gravitational_acceleration = nothing,
                          )

Can anyone help? Some examples of the ShallowWaterModel function would be helpful.
Thanks!

**Comment by navidcy** [comment]
Hi @Lichriszz, could you say what version of Oceananigans.jl you used? E.g. the output of
```julia
julia> using Pkg; Pkg.status()
```
and 
```julia
julia> versioninfo()
```

**Reply by Lichriszz** [reply]
Hi! Thanks for the reply. I am using Oceananigans v0.47.1 and Julia 1.5.3.

julia> using Pkg; Pkg.status()
Status `~/.julia/environments/JuliaPro_v1.5.3-1/Project.toml`
  [c52e3926] Atom v0.12.25 ⚲
  [7073ff75] IJulia v1.23.1
  [033835bb] JLD2 v0.3.3
  [e5e0dc1b] Juno v0.8.4 ⚲
  [9e8cae18] Oceananigans v0.47.1
  [4722fa14] PkgAuthentication v1.1.0
  [91a5bcdd] Plots v1.10.3
  [44d3d7a6] Weave v0.10.6

julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, icelake-client)
Environment:
  JULIA_NUM_THREADS = 4
  JULIA_PKG_SERVER = pkg.juliahub.com
  JULIA_DEPOT_PATH = /Users/lichriszz/.julia:/Applications/JuliaPro-1.5.3-1.app/Contents/Resources/julia/Contents/Resources/julia/local/share/julia:/Applications/JuliaPro-1.5.3-1.app/Contents/Resources/julia/Contents/Resources/julia/share/julia
  JULIA_LOAD_PATH = /Users/lichriszz/.julia/environments/JuliaPro_v1.5.3-1:@:@v#.#:@stdlib

**Reply by navidcy** [reply]
First of all it doesn't seem that you imported `ShallowWaterModel`, e.g., using Oceananigans.ShallowWaterModel`. But also, I'm not sure if that's part of the v0.47.1 release.

Can you give a bit more background on what you are trying to do? E.g., are you trying to reproduce an example from the docs?

**Reply by Lichriszz** [reply]
When I try to do "using Oceananigans.ShallowWaterModel", or using Oceananigans.ShallowWaterModels (with an s), it says shallowwatermodel(s) not defined.
I am not working on a specific problem right now. I am just trying to understand the codes. In the future, we may use the shallow water model to simulate equatorial waves.

**Reply by navidcy** [reply]
May I suggest you try from one of the examples in the Docs?

The ShallowWater model is bot quite “cooked” yet. But it’s getting there!

Other may have more insight on this. @francispoulin, @glwagner

**Reply by Lichriszz** [reply]
Actually we are interested in the shallow water model specifically. But I will go over those examples first. Thanks!

**Reply by navidcy** [reply]
I see you. But unfortunately it's not quite ready. Until you see the ShallowWaterModel appearing in the Documentation then I'd advise not to use it or only "use it at your own risk". :)

**Comment by ali-ramadhan** [comment]
Hi @Lichriszz thanks for reporting this error. As @navidcy said, I think the shallow water model is not yet exported but @francispoulin can better comment on its status.

I've fixed the bug in PR #1363 so it should make it's way into the next tagged release.

If you want to use the shallow water model, for now to get around the bug I think you need to explicitly specify the `architecture`. This version of your initial example worked for me:

```julia
using Oceananigans, Oceananigans.Advection, Oceananigans.Architectures
using Oceananigans.Models: ShallowWaterModel

grid = RegularCartesianGrid(size=(128, 128, 1), extent=(2π, 2π, 2π))

model = ShallowWaterModel(architecture = CPU(), grid = grid, gravitational_acceleration = 10.0)
```

I think you need to specify a number for `gravitational_acceleration ` otherwise you get a different error.

**Reply by Lichriszz** [reply]
Thank you!

**Comment by francispoulin** [comment]
Thanks @Lichriszz for pointing this out and sorry there was a problem to begin with.  

I have played with an example for geostrophic adjustment, both one- and two-dimensional, and now working on the instability of a barotropic Bickley jet.  Coming soon I hope!

What kind of example would you be looking for?

**Reply by Lichriszz** [reply]
Thanks for the reply. I am looking for an example that simulates the evolution of a system, like the incompressible model example here: https://clima.github.io/OceananigansDocumentation/stable/generated/two_dimensional_turbulence/.

**Reply by francispoulin** [reply]
Very good idea!  I'll try and put together an example this afternoon and let you know what I come up with.

**Reply by Lichriszz** [reply]
Thank you!

**Reply by francispoulin** [reply]
@Lichriszz , I created an example and you can find the code at this [link](https://github.com/CliMA/Oceananigans.jl/blob/fjp/two-dimensional-turbulence-shallow-water-example/examples/two_dimensional_turbulence_shallow_water.jl).  It's more of a proof of concept at this stage as the results are not anything special.  Are there particular parameters you want to run?

I also created a PR in case you wanted to make some comments on the code and what make it easier to use, depending on what you want to do with it.

**Reply by Lichriszz** [reply]
Thank you so much. I don't have particular parameters that I want to run.
This example is very helpful.
